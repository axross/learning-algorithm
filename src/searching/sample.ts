export const sortedCharactors = [
  "a",
  "b",
  "c",
  "d",
  "e",
  "f",
  "g",
  "h",
  "i",
  "j",
  "k",
  "l",
  "m",
  "n",
  "o",
  "p",
  "q",
  "r",
  "s",
  "t",
  "u",
  "v",
  "w",
  "x",
  "y",
  "z"
];

export const unsortedCharactors = [
  "c",
  "n",
  "r",
  "j",
  "m",
  "s",
  "b",
  "u",
  "w",
  "y",
  "h",
  "t",
  "v",
  "f",
  "z",
  "g",
  "x",
  "o",
  "k",
  "q",
  "d",
  "e",
  "p",
  "l",
  "a",
  "i"
];
