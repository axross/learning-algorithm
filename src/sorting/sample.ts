import roundRobin from "../test-utility/round-robin.ts";
import randomInteger from "../test-utility/random-integer.ts";

export const charactors = ["a", "b", "c", "d", "e", "f", "g", "h"];

export const samples = roundRobin(charactors).slice(1);

export function getRandomSample() {
  return samples[randomInteger(0, samples.length - 1)];
}

export const staticSample = ["f", "a", "h", "b", "d", "g", "e", "c"];

export const charactorPairs = ["a0", "a1", "b0", "b1", "c0", "c1", "d0", "d1"];

export const charactorPairSamples = roundRobin(["a0", "b0", "c0", "d0"]).reduce(
  (arr, pattern0) => [
    ...arr,
    ...roundRobin(["a1", "b1", "c1", "d1"]).reduce(
      (arr, pattern1) => [...arr, [...pattern0, ...pattern1]],
      [] as string[][]
    )
  ],
  [] as string[][]
);

export const randomCharactorPairSample =
  charactorPairSamples[randomInteger(0, charactorPairSamples.length - 1)];

export const staticCharactorPairSample =
  randomCharactorPairSample[Math.floor((samples.length / 7) * 5)];
