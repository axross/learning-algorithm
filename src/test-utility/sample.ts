function shuffle<Value>(arr: Value[]): Value[] {
  const shuffled = [...arr];

  for (let i = shuffled.length - 1; i >= 0; i--) {
    const rand = Math.floor(Math.random() * (i + 1));

    [shuffled[i], shuffled[rand]] = [shuffled[rand], shuffled[i]];
  }

  return shuffled;
}

export const sampleCharactors = [
  ...Array.from({ length: 5 }, (_, i) => String.fromCharCode(i + 48)),
  ...Array.from({ length: 26 }, (_, i) => String.fromCharCode(i + 65)),
  ...Array.from({ length: 26 }, (_, i) => String.fromCharCode(i + 97))
];

export const shuffledSampleCharactors = shuffle(sampleCharactors);

export const sampleTwoCharactors = [
  ...Array.from({ length: 5 }, (_, i) => [
    String.fromCharCode(i + 48) + "<",
    String.fromCharCode(i + 48) + ">"
  ]),
  ...Array.from({ length: 13 }, (_, i) => [
    String.fromCharCode(i + 65) + "<",
    String.fromCharCode(i + 65) + ">"
  ]),
  ...Array.from({ length: 13 }, (_, i) => [
    String.fromCharCode(i + 97) + "<",
    String.fromCharCode(i + 97) + ">"
  ])
].reduce((array, value) => [...array, ...value], []);

export const shuffledSampleTwoCharactors = [
  ...shuffle([
    ...Array.from({ length: 5 }, (_, i) => String.fromCharCode(i + 48) + "<"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 65) + "<"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 97) + "<")
  ]),
  ...shuffle([
    ...Array.from({ length: 5 }, (_, i) => String.fromCharCode(i + 48) + ">"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 65) + ">"),
    ...Array.from({ length: 13 }, (_, i) => String.fromCharCode(i + 97) + ">")
  ])
];
