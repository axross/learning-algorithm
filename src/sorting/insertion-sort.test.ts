import insertionSort from "./insertion-sort";
import { charactors, getRandomSample, staticSample } from "./sample";

test(`insertionSort() sorts the array to be ordered`, () => {
  for (let i = 1; i <= 100; ++i) {
    const array = [...getRandomSample()];

    insertionSort({
      array,
      compare: (a, b) => a.charCodeAt(0) - b.charCodeAt(0)
    });

    expect(array).toEqual(charactors);
  }
});

describe("insertionSort() sorts the array in the correct process", () => {
  const changesOnSwap: string[][] = [];
  const changesOnInsertion: string[][] = [];
  const array = [...staticSample];

  insertionSort({
    array,
    compare: (a, b) => a.charCodeAt(0) - b.charCodeAt(0),
    onSwap: () => changesOnSwap.push([...array]),
    onInsertion: () => changesOnInsertion.push([...array])
  });

  expect(changesOnSwap).toEqual([
    ["a", "f", "h", "b", "d", "g", "e", "c"],
    ["a", "f", "b", "h", "d", "g", "e", "c"],
    ["a", "b", "f", "h", "d", "g", "e", "c"],
    ["a", "b", "f", "d", "h", "g", "e", "c"],
    ["a", "b", "d", "f", "h", "g", "e", "c"],
    ["a", "b", "d", "f", "g", "h", "e", "c"],
    ["a", "b", "d", "f", "g", "e", "h", "c"],
    ["a", "b", "d", "f", "e", "g", "h", "c"],
    ["a", "b", "d", "e", "f", "g", "h", "c"],
    ["a", "b", "d", "e", "f", "g", "c", "h"],
    ["a", "b", "d", "e", "f", "c", "g", "h"],
    ["a", "b", "d", "e", "c", "f", "g", "h"],
    ["a", "b", "d", "c", "e", "f", "g", "h"],
    ["a", "b", "c", "d", "e", "f", "g", "h"]
  ]);

  expect(changesOnInsertion).toEqual([
    ["a", "f", "h", "b", "d", "g", "e", "c"],
    ["a", "b", "f", "h", "d", "g", "e", "c"],
    ["a", "b", "d", "f", "h", "g", "e", "c"],
    ["a", "b", "d", "f", "g", "h", "e", "c"],
    ["a", "b", "d", "e", "f", "g", "h", "c"],
    ["a", "b", "c", "d", "e", "f", "g", "h"]
  ]);
});
