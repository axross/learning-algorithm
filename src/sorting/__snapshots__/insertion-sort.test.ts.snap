// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the steps of insertionSort() matches the previous snapshot 1`] = `
Object {
  "comparisons": Array [
    Object {
      "a": Object {
        "index": 1,
        "value": "g",
      },
      "b": Object {
        "index": 0,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 2,
        "value": "a",
      },
      "b": Object {
        "index": 0,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 2,
        "value": "a",
      },
      "b": Object {
        "index": 1,
        "value": "g",
      },
    },
    Object {
      "a": Object {
        "index": 3,
        "value": "b",
      },
      "b": Object {
        "index": 0,
        "value": "a",
      },
    },
    Object {
      "a": Object {
        "index": 3,
        "value": "b",
      },
      "b": Object {
        "index": 1,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 3,
        "value": "b",
      },
      "b": Object {
        "index": 2,
        "value": "g",
      },
    },
    Object {
      "a": Object {
        "index": 4,
        "value": "c",
      },
      "b": Object {
        "index": 0,
        "value": "a",
      },
    },
    Object {
      "a": Object {
        "index": 4,
        "value": "c",
      },
      "b": Object {
        "index": 1,
        "value": "b",
      },
    },
    Object {
      "a": Object {
        "index": 4,
        "value": "c",
      },
      "b": Object {
        "index": 2,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 4,
        "value": "c",
      },
      "b": Object {
        "index": 3,
        "value": "g",
      },
    },
    Object {
      "a": Object {
        "index": 5,
        "value": "d",
      },
      "b": Object {
        "index": 0,
        "value": "a",
      },
    },
    Object {
      "a": Object {
        "index": 5,
        "value": "d",
      },
      "b": Object {
        "index": 1,
        "value": "b",
      },
    },
    Object {
      "a": Object {
        "index": 5,
        "value": "d",
      },
      "b": Object {
        "index": 2,
        "value": "c",
      },
    },
    Object {
      "a": Object {
        "index": 5,
        "value": "d",
      },
      "b": Object {
        "index": 3,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 5,
        "value": "d",
      },
      "b": Object {
        "index": 4,
        "value": "g",
      },
    },
    Object {
      "a": Object {
        "index": 6,
        "value": "e",
      },
      "b": Object {
        "index": 0,
        "value": "a",
      },
    },
    Object {
      "a": Object {
        "index": 6,
        "value": "e",
      },
      "b": Object {
        "index": 1,
        "value": "b",
      },
    },
    Object {
      "a": Object {
        "index": 6,
        "value": "e",
      },
      "b": Object {
        "index": 2,
        "value": "c",
      },
    },
    Object {
      "a": Object {
        "index": 6,
        "value": "e",
      },
      "b": Object {
        "index": 3,
        "value": "d",
      },
    },
    Object {
      "a": Object {
        "index": 6,
        "value": "e",
      },
      "b": Object {
        "index": 4,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 6,
        "value": "e",
      },
      "b": Object {
        "index": 5,
        "value": "g",
      },
    },
    Object {
      "a": Object {
        "index": 7,
        "value": "h",
      },
      "b": Object {
        "index": 0,
        "value": "a",
      },
    },
    Object {
      "a": Object {
        "index": 7,
        "value": "h",
      },
      "b": Object {
        "index": 1,
        "value": "b",
      },
    },
    Object {
      "a": Object {
        "index": 7,
        "value": "h",
      },
      "b": Object {
        "index": 2,
        "value": "c",
      },
    },
    Object {
      "a": Object {
        "index": 7,
        "value": "h",
      },
      "b": Object {
        "index": 3,
        "value": "d",
      },
    },
    Object {
      "a": Object {
        "index": 7,
        "value": "h",
      },
      "b": Object {
        "index": 4,
        "value": "e",
      },
    },
    Object {
      "a": Object {
        "index": 7,
        "value": "h",
      },
      "b": Object {
        "index": 5,
        "value": "f",
      },
    },
    Object {
      "a": Object {
        "index": 7,
        "value": "h",
      },
      "b": Object {
        "index": 6,
        "value": "g",
      },
    },
  ],
  "from": Array [
    "f",
    "g",
    "a",
    "b",
    "c",
    "d",
    "e",
    "h",
  ],
  "insertions": Array [
    Object {
      "before": Object {
        "index": 1,
        "value": "g",
      },
      "chosen": Object {
        "index": 1,
        "value": "g",
      },
    },
    Object {
      "before": Object {
        "index": 0,
        "value": "f",
      },
      "chosen": Object {
        "index": 2,
        "value": "a",
      },
    },
    Object {
      "before": Object {
        "index": 1,
        "value": "f",
      },
      "chosen": Object {
        "index": 3,
        "value": "b",
      },
    },
    Object {
      "before": Object {
        "index": 2,
        "value": "f",
      },
      "chosen": Object {
        "index": 4,
        "value": "c",
      },
    },
    Object {
      "before": Object {
        "index": 3,
        "value": "f",
      },
      "chosen": Object {
        "index": 5,
        "value": "d",
      },
    },
    Object {
      "before": Object {
        "index": 4,
        "value": "f",
      },
      "chosen": Object {
        "index": 6,
        "value": "e",
      },
    },
    Object {
      "before": Object {
        "index": 7,
        "value": "h",
      },
      "chosen": Object {
        "index": 7,
        "value": "h",
      },
    },
  ],
  "to": Array [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
    "g",
    "h",
  ],
}
`;
